src/Symfony/Component/Runtime/autoload.php

```diff
@@ -0,0 +1,32 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+use Symfony\Component\Runtime\SymfonyRuntime;
+
+if (true === require_once dirname(__DIR__, 2).'/autoload.php') {
+    return;
+}
+
+if (!isset($_SERVER['APP_RUNTIME'])) {
+    $runtime = new SymfonyRuntime($_SERVER['APP_RUNTIME_OPTIONS'] ?? []);
+} elseif (is_string($runtime = $_SERVER['APP_RUNTIME'])) {
+    $runtime = new $runtime($_SERVER['APP_RUNTIME_OPTIONS'] ?? []);
+}
+
+if (1 === $main = require $_SERVER['SCRIPT_FILENAME']) {
+    exit(0);
+}
+
+[$main, $arguments] = $runtime->resolveMain($main);
+$result = $main(...$arguments);
+$handler = $runtime->getHandler($result);
```

Comment:
Conceptual quesiton. Shouldn't this one be the part of `Main`? [Here](https://github.com/symfony/symfony/pull/36652#discussion_r418938697) you referencing Main as something like C, Rust. I suppose Java also goes here. Mains in these langs do not make users to interpret the result, they just return the exit code, I hope

So I (as user) expect that after executing main I have nothing to do - program did it's job. 

What if we just move result processing inside of Main closures? There will be more cohesion, since each result processing routing will know exactly how this result was obtained and there will be no need in type guessing of result later.

With current approach there is also a problem - what if I want handle same result types of different Mains differently?

Reply:
> move result processing inside of Main closures

That would defeat the purpose of this proposal, which is also about decoupling the return value from the runtime. E.g. this PR allows that returned `Kernel` to be run in PHP-FPM and/or in PHP-PM indifferently (provided the app doesn't use global state like PHP sessions of course).

> what if I want handle same result types of different Mains differently?

That would be the responsibility of the runtime. To do so, the runtime would decorate the "main" to return a wrapped result, which would contain the processing instructions.

Reply:
But current implementation makes in some cases actual application to run outside the Main code. I think this is not that expected by the end user.

https://github.com/symfony/symfony/blob/0c09b96c03ca051904350eca809dcb1983269b3a/src/Symfony/Component/Runtime/SymfonyRuntime.php#L88-L97

So IIUC the all actual code of application will happen in `exit($handler($result))`

Reply:
I don't get your point sorry. If it's about matching supposed expectations to actual behavior, this looks like numerology to me...

Anything more concrete for thoughts? To me, all is as I wish they were :)

Reply:
My point that the `Main` is actually not a `Main` in current implementation. It looks like a `MainFactory` at most cases. This function **does not run** the application code, which `Main` in all other languages do

Reply:
OK, so naming things... :) Any proposal?

Reply:
Let's start with straight and obvious naming, trying to keep your ideas. IIUC current code works like this

```
[$resolvedMainFactory, $arguments] = $runtime->resolveMainFactory($mainFactory);
$main = $resolvedMainFactory(...$arguments);
$handler = $runtime->getHandler($main);

exit($handler($main));
```

We can make it more simple by moving actual call of `$mainFactory(...$arguments)` making it really resolve main again

```
$main = $runtime->resolveMain($mainFactory); // does internal call of $resolvedMainFactory(...$arguments) and hiding the arguments routine
$handler = $runtime->getHandler($main);

exit($handler($main));
```

In this case `Main` could be anything - sf kernel, sf console app or other "invokable" here. I think `Executor` suits more here. So the final code might look like

```php
// I skip some setup below
$runtime = new SymfonyRuntime($_SERVER['APP_RUNTIME_OPTIONS'] ?? []);
if (1 === $mainFactory = require $_SERVER['SCRIPT_FILENAME']) {
    exit(0);
}

$main = $runtime->resolveMain($mainFactory); // or createMain? 
$executor = $runtime->getExecutor($main);

exit($executor($main));
```

I hope this make code a bit clearer where the job is done and how it's prepared fo the run

Reply:
> // does internal call of $resolvedMainFactory(...$arguments) and hiding the arguments routine

rejected: this is breaking all benefits of the current design (the "high-order" part, which is the one that provides extensibility.)

Sticking to naming, another idea:
```php
if (1 === $app = require $_SERVER['SCRIPT_FILENAME']) {
    exit(0);
}

[$app, $arguments] = $runtime->resolveApp($app);
$result = $app(...$arguments);
$main = $runtime->getMain($result);

exit($main($result));
```

/cc @Nyholm since you wondered about this aspect too.

Reply:
Looks better, but still has a problem with `result`. Calling "main" function on "result" make non sense for me. Dropping the rejected idea my suggestion would be

```
[$mainFactory, $arguments] = $runtime->resolveMainFactory($mainFactory);
$main = $mainFactory(...$arguments);
$executor = $runtime->getExecutor($main);

exit($exectutor($main));
```

From my POV your `result` is the real `main` - the place where the application runs, so I suggest to use it as `main`, not as `result`.

Also I do not see benefits from introducing `app` here, since it's not an app in any case and this `app` lives only two lines of code, serving like `Main` factory

Reply:
Another try:
```php
[$starter, $arguments] = $runtime->resolveStarter($starter);
$app = $starter(...$arguments);
$main = $runtime->getMain($app);
```

Reply:
IIUC this result in final invokation like `exit($main($app))`. Well, it has more sense, we got an advance. In this case - what about not sticking to `Main` term at all? I think a term designated to code invokation may have clearer meaning, i.e
```php
$executor = $runtime->getExecutor($app);
exit($executor($app));
$runner = $runtime->getRunner($app);
exit($runner($app));
$handler = $runtime->getHandler($app);
exit($handler($app));
```

Reply:
Thanks to this discussion, I found a better design (not only naming anymore):
```
$app = $runtime->getApp($app)();

exit($runtime->getMain($app)());
```

Reply:
I reviewed this. I like this name better. The only thing that is a bit strange is that $app may contain a Response. But I’m fine with that :)

Reply:
Yay, and you had also nicely hidden the $arguments calling part. I love it. New namings are more clear too. The usage is much more obvious with these improvements!